src/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Ssl.cs

```diff
@@ -569,6 +577,60 @@ internal static void SslSetTargetName(SafeSslHandle sslHandle, string targetName
             throw new SslException();
         }
 
+        internal static unsafe void SslCtxSetAlpnProtos(SafeSslHandle ctx, List<SslApplicationProtocol> protocols)
+        {
+            byte[] buffer = ConvertAlpnProtocolListToByteArray(protocols);
+            int osStatus;
+
+            fixed (byte* b = buffer)
+            {
+                int result = SslCtxSetAlpnProtos(ctx, (IntPtr)b, buffer.Length, out osStatus);
+                if (result != 1)
+                {
+                    throw CreateExceptionForOSStatus(osStatus);
+                }
+            }
+        }
+
+        internal static byte[] SslGetAlpnSelected(SafeSslHandle ssl)
+        {
+            IntPtr protocol;
+            int len;
+            SslGetAlpnSelected(ssl, out protocol, out len);
+
+            if (len == 0)
+                return null;
+
+            byte[] result = new byte[len];
+            Marshal.Copy(protocol, result, 0, len);
+            return result;
+        }
+
+        internal static byte[] ConvertAlpnProtocolListToByteArray(List<SslApplicationProtocol> applicationProtocols)
+        {
+            int protocolSize = 0;
+            foreach (SslApplicationProtocol protocol in applicationProtocols)
+            {
+                if (protocol.Protocol.Length == 0 || protocol.Protocol.Length > byte.MaxValue)
+                {
+                    throw new ArgumentException(SR.net_ssl_app_protocols_invalid, nameof(applicationProtocols));
+                }
+
+                protocolSize += protocol.Protocol.Length + 1;
+            }
+
+            byte[] buffer = new byte[protocolSize];
+            var offset = 0;
+            foreach (SslApplicationProtocol protocol in applicationProtocols)
+            {
+                buffer[offset++] = (byte)(protocol.Protocol.Length);
+                protocol.Protocol.Span.CopyTo(new Span<byte>(buffer).Slice(offset));
+                offset += protocol.Protocol.Length;
```

Comment:
Rather than continually switching to a Span it seems like what you want is something like

```C#
byte[] buffer = new byte[protocolSize];
Span<byte> dest = buffer;

foreach (SslApplicationProtocol protocol in applicationProtocols)
{
    dest[0] = (byte)protocol.Protocol.Length;
    dest = dest.Slice(1);

    protocol.Protocol.Span.CopyTo(dest);
    dest = dest.Slice(protocol.Protocol.Length);
}

return buffer;
```

Alternatively, keeping the code mostly as it is (with a tracked offset) but saving the Span outside the loop.